hack/verify-shellcheck.sh

```diff
@@ -88,40 +110,46 @@ array_contains () {
   return $in
 }
 
-# creates the shellcheck container for later user
-create_container () {
-  # TODO(bentheelder): this is a performance hack, we create the container with
-  # a sleep MAX_INT32 so that it is effectively paused.
-  # We then repeatedly exec to it to run each shellcheck, and later rm it when
-  # we're done.
-  # This is incredibly much faster than creating a container for each shellcheck
-  # call ...
-  docker run --name "${SHELLCHECK_CONTAINER}" -d --rm -v "${KUBE_ROOT}:${KUBE_ROOT}" -w "${KUBE_ROOT}" --entrypoint="sleep" "${SHELLCHECK_IMAGE}" 2147483647
-}
-# removes the shellcheck container
-remove_container () {
-  docker rm -f "${SHELLCHECK_CONTAINER}" &> /dev/null || true
-}
+# detect if the host machine has the required shellcheck version installed
+# if so, we will use that instead.
+HAVE_SHELLCHECK=false
+if which shellcheck &>/dev/null; then
+  detected_version="$(shellcheck --version | grep 'version: .*')"
+  if [[ "${detected_version}" = "version: ${SHELLCHECK_VERSION}" ]]; then
+    HAVE_SHELLCHECK=true
+  fi
+fi
 
-# remove any previous container, ensure we will attempt to cleanup on exit,
-# and create the container
-remove_container
-trap remove_container EXIT
-if ! output="$(create_container 2>&1)"; then
-    {
-      echo "Failed to create shellcheck container with output: "
-      echo ""
-      echo "${output}"
-    } >&2
-    exit 1
+# tell the user which we've selected and possibly set up the container
+if $HAVE_SHELLCHECK; then
+  echo "Using host shellcheck ${SHELLCHECK_VERSION} binary."
+else
+  echo "Using shellcheck ${SHELLCHECK_VERSION} docker image."
+  # remove any previous container, ensure we will attempt to cleanup on exit,
+  # and create the container
+  remove_container
+  trap remove_container EXIT
```

Comment:
we should be using `kube::util::trap_add`

Reply:
fair. doing that. I'd like to make a version of this script useable by other repos without importing the whole of hack/ at some point, but I'll deal with that then.

Reply:
done